### DISCUSSION TITLE
can duplicated manual declaration of properties, collections be avoided in the following scenario?

### DISCUSSION THREAD

**Discussion by shimpe** [AvaloniaUI/Avalonia#16797] [Q&A] [answered]
Using MVVM, imagine a user control that contains a search bar and a DataGrid. To populate the DataGrid (using dynamic data) in my MainWindowViewModel I added a SourceCache and a ReadOnlyObservableCollection. For the search bar I have a SearchText property. I can bind to all these without problem - everything works as expected.

Now imagine I have a TabControl containing 16 TabItems, each of which contain an (independent of the other tabs) search bar and DataGrid. Suddenly my MainWindowViewModel requires 16 SearchText (SearchText1, SearchText2, ..., SearchText16) properties, 16 sourceCaches, and 16 ReadOnlyObservableCollections.

The fact that I need 16 of each is expected, but is there a way to create/contain all these SearchText properties, caches and collections in a list/array/..? and bind to list/array elements because the number of properties and collections I now have to declare, initialize, bind, ... manually is exploding and the duplication feels wrong.

Maybe I'm going about this in completely the wrong way and there's a more appropriate way of handling such situation? If so, how would an experienced MVVM/AvaloniaUI developer handle this? Needless to say, I'm still quite new to csharp and MVVM/AvaloniaUI.

**Comment by workgroupengineering** [comment, marked as answer]
pseudo code

```c#
public class ItemDettailVM:ViewModelBase
{
    public string?  SearchText {get;set}
    
    public ReadOnlyObservableCollection<TResut> Result {get;}
}

public MainWindowViewModel: ViewModelBase
{
  public ReadOnlyObservableCollection<ItemDettailVM> Items {get;}
}
```

```xaml
<Window ...
    xmlns:vm="" 
    x:DataType="vm:MainWindowViewModel">
  <Window.DataContex>
     <vm:MainWindowViewModel/>
  </Window.DataContex>
  <TabControl ItemsSource="{Binding Items }">
      <TabControl.ItemTemplate>
        <DataTemplate DataType="vm:ItemDettailVM">
            <StackPanel>
                 <TextBox Text="{Binding SearchText}"/>
                 <DataGrid ItemsSource="{Binding Result }"/>
             </StackPanel>
       </DataTemplate>
     <TabControl.ItemTemplate>
  </TabControl >
</Window> 
```

**Comment by shimpe** [comment]
(I guess TabControl.ItemTemplate should have been TabControl.ContentTemplate?)

Ah great! That's indeed what I'm looking for. 

I do have 2 tabs with different content, but I guess I can create a data template selector based on this example: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/DataTemplates/IDataTemplateSample

Thanks!
